pkg/querier/memory_tracker_queryable.go

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querier
+
+import (
+	"context"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+// memoryTrackerSampleAndChunkQueryable wraps a storage.SampleAndChunkQueryable and ensures all queries have an unlimited
+// MemoryConsumptionTracker in their context.
+type memoryTrackerSampleAndChunkQueryable struct {
+	inner storage.SampleAndChunkQueryable
+}
+
+// NewMemoryTrackerSampleAndChunkQueryable wraps a storage.Queryable to inject an unlimited MemoryConsumptionTracker
+// into the context for all queriers.
+func NewMemoryTrackerSampleAndChunkQueryable(inner storage.Queryable) storage.Queryable {
+	// Try to cast to SampleAndChunkQueryable to preserve the interface
+	if scq, ok := inner.(storage.SampleAndChunkQueryable); ok {
+		return &memoryTrackerSampleAndChunkQueryable{inner: scq}
+	}
+	// Fall back to wrapping just as Queryable
+	return &memoryTrackerQueryable{inner: inner}
+}
+
+func (q *memoryTrackerSampleAndChunkQueryable) Querier(minT, maxT int64) (storage.Querier, error) {
+	querier, err := q.inner.Querier(minT, maxT)
+	if err != nil {
+		return nil, err
+	}
+	return &memoryTrackerQuerier{inner: querier}, nil
+}
+
+func (q *memoryTrackerSampleAndChunkQueryable) ChunkQuerier(minT, maxT int64) (storage.ChunkQuerier, error) {
+	querier, err := q.inner.ChunkQuerier(minT, maxT)
+	if err != nil {
+		return nil, err
+	}
+	return &memoryTrackerChunkQuerier{inner: querier}, nil
+}
+
+// memoryTrackerQueryable is for queryables that don't support ChunkQuerier
+type memoryTrackerQueryable struct {
+	inner storage.Queryable
+}
+
+func (q *memoryTrackerQueryable) Querier(minT, maxT int64) (storage.Querier, error) {
+	querier, err := q.inner.Querier(minT, maxT)
+	if err != nil {
+		return nil, err
+	}
+	return &memoryTrackerQuerier{inner: querier}, nil
+}
+
+// memoryTrackerQuerier wraps a storage.Querier and ensures all Select calls have an unlimited
+// MemoryConsumptionTracker in their context.
+type memoryTrackerQuerier struct {
+	inner storage.Querier
+}
+
+func (q *memoryTrackerQuerier) Select(ctx context.Context, sortSeries bool, hints *storage.SelectHints, matchers ...*labels.Matcher) storage.SeriesSet {
+	if _, err := limiter.MemoryConsumptionTrackerFromContext(ctx); err != nil {
+		ctx = limiter.ContextWithNewUnlimitedMemoryConsumptionTracker(ctx)
```

Comment:
Could you please add a comment here explaining why we don't unconditionally add an unlimited memory consumption tracker?

Reply:
Actually I think we don't need this conditional injection. We started this tracker from the place that we know has no memory tracker, from initRuler.

Removed the conditional injection in 38f02e1e1048b9d4cd5feadf26cf82cd5a486a1a

Let me know if you oversees that the conditional injection is actually necessary.

Reply:
If the ruler is running without a query-frontend (ie. is evaluating queries inside the ruler process, rather than delegating to a query-frontend), then won't it use this `Queryable` / `Querier` with the MQE instance evaluating queries?